### DISCUSSION TITLE
Can't set target temperature of freezer

### DISCUSSION THREAD

**Discussion by Pantalooney** [astrandb/miele#481] [Q&A]
I have a new Miele freezer and have managed to include it in HA using this integration. Many kudos to @astrandb for this excellent work, this integration is amazing!

However, I struggle to set the target temperature of the freezer. There is a sensor "freezer.zieltemperatur", which displays the current value, but I can't find a way to change it. 

The WIki even mentions setting the freezer target temperature, so there must be a way to do it. 

I am not new to HA, but still very from from being an expert - so maybe I need a hint where to change what in which yaml ... ?

**Comment by Pantalooney** [comment]
Still would be very interested to learn how I could make this work. 
The target temperature can be set through the Miele app, so I'm confident there is a way to do it in Home Assistant as well. 

Also, I got the following response from GET ACTIONS:
```
{
  "processAction": [
    4
  ],
  "light": [],
  "ambientLight": [],
  "startTime": [],
  "ventilationStep": [],
  "programId": [],
  "targetTemperature": [
    {
      "zone": 1,
      "min": -26,
      "max": -16
    }
  ],
  "deviceName": true,
  "powerOn": false,
  "powerOff": false,
  "colors": [],
  "modes": [
    1
  ],
  "runOnTime": []
}`
```

To my very limited understanding this means that it can be set.

Anyone got an idea how?

**Reply by astrandb** [reply]
The sensor freezer.zieltemperatur is indeed a readonly sensor.
In the UI you just bring up the climate entity for the freezer and adjust the dial. 
If you want to set it in an automation or a script, then goto Developer tools->Actions, select Climate: Set target temperature. Choose the target device and set desired target temperature.  Then press Perform action.

**Comment by Pantalooney** [comment]
Hi @astrandb, many many thanks for the reply!

Unfortunately, I had already tried that, to no avail. I can jog the dial for climate.freezer_gefrierschrank (in the screenshot below, it is .17 °C and I changed it to 19 °C), but it has no effect on the target temperature. 

![image](https://github.com/user-attachments/assets/5f396711-0a07-48d5-bc4e-8f32b043d2a5)
![image](https://github.com/user-attachments/assets/b8f2151b-55b9-4e4c-90f1-db5074bb317e)

**Comment by Pantalooney** [comment]
One more observation, maybe this helps. When I try to set the target temperature in an automation, I cannot enter negative values. It says the value has to be 0 or larger.

Apparently it accepts only values between 0 and 250 °C. Is this configured somewhere?

**Comment by astrandb** [comment]
It works perfectly with my freezer. I can change on the climate entity or with an automation action and it is reflected within less than a second in the IOS app - and vice versa.

The Number selector in the automation editor is controlled by HA and not part of this integration. The upper and lower limits have been respected before as far as I can remember. However, the limits are not working here either. But I can type -19, and it is posted with expected result.

If you generate a diagnostic file from the device page and mail it to me ake@stramdnerg.eu I can have a look that your freezer behaves as expected in the API.

Regarding the non-functioning limits in the number selector I recommend that you look for similar issue reports in HA core and frontend repos and creates a report if you are the first to find this problem.

**Comment by Pantalooney** [comment]
Thanks, I have sent the file. Btw, I can add "-19" to the automation by editing the yaml, and when I execute it, it says it was executed successfully, but the target temperature still does not change.

It's not a problem to set other settings (Superfrost on/off). 

In parallel I will check for a bug fix with HA core about the UI not allowing negative values.

**Reply by astrandb** [reply]
It sounds as a problem in Miele's end. I cannot find any surprises in the diagnostic file.
If you are up to some serious debugging you can turn on debug logging (On the integration card). Turn on-> change target temperature on the climate entity-> turn off debugging and examine the file for warnings or errors when sending the command.
Another interesting exercise is to use https://www.miele.com/developer/swagger-ui/ . Click authorize and enter your credentials. You can the play around with the API and see if you can reproduce the problem or if it just works as expected.

**Comment by Pantalooney** [comment]
Many thanks! Sorry, I did not find the time to look into it the past days. 
I've tried debugging, but cannot find anything suspicious. All that it logs when I try to change the value is the following:
```
2024-12-15 16:37:34.684 DEBUG (MainThread) [custom_components.miele.climate] kwargs: {'entity_id': ['climate.freezer_gefrierschrank'], 'temperature': -20.0}
2024-12-15 16:37:34.734 DEBUG (MainThread) [pymiele.pymiele] set_target res: 204
2024-12-15 16:37:39.290 DEBUG (MainThread) [custom_components.miele] Finished fetching miele data in 0.042 seconds (success: True)
```
I had already stumbeled upon the swagger UI the other day, but couldn't really make sense of it in my then limited time. I have now spent some more time on it, and if I interpret the results correctly, you might be right about a Miele problem, however, I am wondering about the "Malformed request" return statement. 

I am using the PUT ACTION section to send the following code:
```
{
  "targetTemperature": [
    {
      "zone": 1,
      "value": -20
    }
  ]
}
```
Responses:

Curl (confidential IDs replaced with xxxxx):
```
curl` -X 'PUT' \
  'https://api.mcs3.miele.com/v1/devices/0007124xxxxx/actions' \
  -H 'accept: */*' \
  -H 'Authorization: Bearer DE_xxxxx' \
  -H 'Content-Type: application/json' \
  -d '{
  "targetTemperature": [
    {
      "zone": 1,
      "value": -20
    }
  ]
}
```
and then:
204 | No content | No links
400 | Malformed requestMedia typeapplication/jsonExample ValueSchema{   "code": 400,   "message": "Malformed request" } | No links
401 | Authorization failedMedia typeapplication/jsonExample ValueSchema{   "code": 401,   "message": "Unauthorized" } | No links
404 | Not foundMedia typeapplication/jsonExample ValueSchema{   "code": 404,   "message": "HTTP 404 Not found" } | No links
405 | Method not allowedMedia typeapplication/jsonExample ValueSchema{   "code": 405,   "message": "HTTP 405 Method Not Allowed" } | No links
500 | Internal Server ErrorMedia typeapplication/jsonExample ValueSchema{   "message": "Internal Server Error" } | No links
503 | unavailableMedia typeapplication/jsonExample ValueSchema{  "code": 503,  "message": "unavailable" }

Any idea where this "malformed request" might come from?

**Reply by astrandb** [reply]
> 2024-12-15 16:37:34.734 DEBUG (MainThread) [pymiele.pymiele] set_target res: 204

You get a 204 which means that the request was accepted. 204-503 is just a list of possible return codes.

The conclusion is that the Miele API just swallows the targetTemperature action without acting on it.

**Reply by Pantalooney** [reply]
Many thanks! Seems I have to submit a bug report to Miele then.

**Reply by joel-bourquard** [reply]
It's strange: I have a similar problem but for me the non-working target temperature adjustment (ie: HTTP 204 with no effects) is happening with a fridge (model "K 4323 DD ws").

With my freezer (model "FN 27474 ws") it works perfectly.
